apps/mark-scan/backend/src/custom-paper-handler/state_machine.ts

```diff
@@ -1,267 +1,394 @@
+    // so we wait to decrease chance of failed load
+    setTimeout(async () => {
+      await context.driver.loadPaper();
+      await context.driver.parkPaper();
+    }, 300);
+  },
+  on: {
+    PAPER_INSIDE: {
+      target: 'waiting_for_ballot_data',
+    },
+  },
+};
 
-  async scanAndSave(pathOut: string): Promise<void> {
-    await this.driver.setScanDirection('backward');
-    const grayscaleResult = await this.driver.scan();
-    const grayscaleData = grayscaleResult.data;
-    const colorResult = createImageData(
-      grayscaleResult.width,
-      grayscaleResult.height
-    );
-    const colorData = new Uint32Array(
-      colorResult.data.buffer,
-      colorResult.data.byteOffset,
-      colorResult.data.byteLength
-    );
-    for (let i = 0; i < grayscaleData.byteLength; i += 1) {
-      const luminance = grayscaleData[i];
-      colorData[i] =
-        // eslint-disable-next-line no-bitwise
-        (luminance << 24) | (luminance << 16) | (luminance << 8) | 255;
-    }
-    await writeImageData(pathOut, colorResult);
-  }
+const WaitingForBallotDataState: StateNodeConfig<
+  Context,
+  any,
+  PaperHandlerStatusEvent,
+  BaseActionObject
+> = {
+  invoke: pollPaperStatus(),
+  on: {
+    VOTER_INITIATED_PRINT: {
+      target: 'printing_ballot',
+    },
+  },
+};
 
-  async printBallot(
-    pdfData: Uint8Array,
-    options: Partial<ImageConversionOptions> = {}
-  ): Promise<void> {
-    this.status = 'printing_ballot';
-    const enablePrintPromise = this.enablePrint();
+const PrintingBallotState: StateNodeConfig<
+  Context,
+  any,
+  PaperHandlerStatusEvent,
+  BaseActionObject
+> = {
+  invoke: pollPaperStatus(),
+  on: {
+    BALLOT_PRINTED: {
+      target: 'scanning',
+    },
+  },
+};
 
-    let time = Date.now();
-    const pages: ImageData[] = [];
-    for await (const { page, pageCount } of pdfToImages(Buffer.from(pdfData), {
-      scale: 200 / 72,
-    })) {
-      assert(pageCount === 1, `Unexpected page count ${pageCount}`);
-      pages.push(page);
-    }
-    const page = pages[0];
-    assert(page, 'Unexpected undefined page');
-    debug(`pdf to image took ${Date.now() - time} ms`);
-    time = Date.now();
+const ScanningState: StateNodeConfig<
+  Context,
+  any,
+  PaperHandlerStatusEvent,
+  BaseActionObject
+> = {
+  invoke: {
+    id: 'scanAndSave',
+    src: (context) => {
+      return scanAndSave(context.driver);
+    },
+    onDone: {
+      target: 'interpreting',
+      actions: assign({ scannedImagePaths: (_, event) => event.data }),
+    },
+  },
+};
 
-    const ballotBinaryBitmap = imageDataToBinaryBitmap(page, options);
-    debug(`bitmap width: ${ballotBinaryBitmap.width}`);
-    debug(`bitmap height: ${ballotBinaryBitmap.height}`);
-    debug(`image to binary took ${Date.now() - time} ms`);
-    time = Date.now();
+const InterpretingState: StateNodeConfig<
+  Context,
+  any,
+  PaperHandlerStatusEvent,
+  BaseActionObject
+> = {
+  invoke: [
+    { ...pollPaperStatus() },
```

Comment:
Do we need to poll paper status here? I don't see any transitions out of this state based on paper status.

Reply:
`paperStatus` is a bit of a misnomer; it should be something like `pollHardwareStatus`. I've been leaving this in every step because in the future we'll need to handle changes to paper and hardware status. But like you note in `WaitingForBallotDataState` it's not needed right now and this step is at lower risk of abnormal hardware status, so I'll remove as well.